### PR TITLE
Added the flat c layer for the achievements api

### PR DESCRIPTION
**Pull request by rakhad**

### CODE REVIEW

**Comment by jplafonta** [comment]
>       <AdditionalIncludeDirectories>$(ProjectDir)..\..\Source\Services\TitleStorage;$(ProjectDir);$(ProjectDir)..\..\Source\Services\Common;$(ProjectDir)..\..\Source\Services\Social;$(ProjectDir)..\..\Source\Services\Multiplayer;$(ProjectDir)..\..\Source\Services\Presence;$(ProjectDir)..\..\Source\Services\Misc;$(ProjectDir)..\..\Source\Services\Social\Manager;$(ProjectDir)..\..\Source\Services\RealTimeActivity;$(ProjectDir)..\..\Source\Shared;$(ProjectDir)..\..\Source\Shared\WinRT;$(ProjectDir)..\..\Include;$(ProjectDir)..\..\Source\System;$(ProjectDir)..\..\Source\Services\Leaderboard;$(ProjectDir)..\..\Source\Services\Stats\Manager;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>

Add ..\..\Source\Services\Achievements to additional include directories

---
Refers to: CppSource/Build/Microsoft.Xbox.Services.140.UWP.C/Microsoft.Xbox.Services.140.UWP.C.vcxproj:97 in 1211885. [](commit_id = 12118853980a0de9685b9445b1124298087e60bb, deletion_comment = False)

File: CppSource/Include/xsapi/achievements_c.h [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    bool isSecret;
+    PCSTR unlockedDescription;
+    PCSTR lockedDescription;
+    PCSTR productId;
+    XSAPI_ACHIEVEMENT_TYPE type;
+    XSAPI_ACHIEVEMENT_PARTICIPATION_TYPE participationType;
+    XSAPI_ACHIEVEMENT_TIME_WINDOW* available;
+    XSAPI_ACHIEVEMENT_REWARD** rewards;
+    uint32_t rewardsCount;
+    uint64_t estimatedUnlockTime;
```
Comment (jplafonta) [inline CppSource/Include/xsapi/achievements_c.h]: is this not time_t?

File: CppSource/Include/xsapi/achievements_c.h [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    uint32_t itemsCount;
+    bool hasNext;
+
+    XSAPI_ACHIEVEMENTS_RESULT_IMPL* pImpl;
+} XSAPI_ACHIEVEMENTS_RESULT;
+
+XSAPI_DLLEXPORT bool XBL_CALLING_CONV
+AchievementsResultHasNext(
+    _In_ XSAPI_ACHIEVEMENTS_RESULT* achievementsResult
+) XSAPI_NOEXCEPT;
```
Comment (jplafonta) [inline CppSource/Include/xsapi/achievements_c.h]: nit: indent here and elsewhere

File: CppSource/Include/xsapi/achievements_c.h [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+typedef void(*XSAPI_UPDATE_ACVHIEVEMENT_COMPLETION_ROUTINE)(
+    _In_ XSAPI_RESULT_INFO result,
+    _In_opt_ void* context
+    );
+XSAPI_DLLEXPORT XSAPI_RESULT XBL_CALLING_CONV
+AchievementServiceUpdateAchievement(
+    _In_ XSAPI_XBOX_LIVE_CONTEXT* pContext,
+    _In_ PCSTR xboxUserId,
+    _In_opt_ uint32_t titleId,
```
Comment (jplafonta) [inline CppSource/Include/xsapi/achievements_c.h]: is this actually optional? if so should it be uint32_t*?

File: CppSource/Include/xsapi/achievements_c.h (hunk 1 above)
Comment (rakhad) [reply CppSource/Include/xsapi/achievements_c.h]: No it is type const std::chrono::seconds& which is a uint64

File: CppSource/Source/Services/Achievements/achievements.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+CATCH_RETURN()
+
+HC_RESULT get_next_execute(
+    _In_opt_ void* context,
+    _In_ HC_TASK_HANDLE taskHandle
+)
+{
+    auto args = reinterpret_cast<get_next_taskargs*>(context);
+    xbox_live_result<achievements_result> result = args->achievementsResult->pImpl->cppAchievementsResult().get_next(args->maxItems).get();
+    args->copy_xbox_live_result(result);
```
Comment (jplafonta) [inline CppSource/Source/Services/Achievements/achievements.cpp]: it seems like the achievement result doesn't get passed back, only the error message

File: CppSource/Include/xsapi/achievements_c.h [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    _In_ uint64_t taskGroupId
+    ) XSAPI_NOEXCEPT;
+
+typedef void(*XSAPI_GET_ACHIEVEMENTS_FOR_TITLE_ID_COMPLETION_ROUTINE)(
+    _In_ XSAPI_RESULT_INFO result,
+    XSAPI_ACHIEVEMENTS_RESULT* achievementsResult,
+    _In_opt_ void* context
+    );
+XSAPI_DLLEXPORT XSAPI_RESULT XBL_CALLING_CONV
+AchievementServiceGetAchievementsForTitleId(
```
Comment (jplafonta) [inline CppSource/Include/xsapi/achievements_c.h]: AchievementService* is a little inconsistent with the naming I have been using.  For example for Title storage my APIs are just TitleStorage* and not TitleStorageService*.  Its fine for now but when we finalize these API's we should be consistent on the naming conventions.

File: CppSource/Source/Services/Achievements/achievements_helper.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    auto achievement = new XSAPI_ACHIEVEMENT();
+    achievement->pImpl = new XSAPI_ACHIEVEMENT_IMPL(cppAchievement, achievement);
+    return achievement;
+}
+
+XSAPI_ACHIEVEMENTS_RESULT* CreateAchievementsResultFromCpp(
+    _In_ achievements_result cppResult
+)
+{
+    auto result = new XSAPI_ACHIEVEMENTS_RESULT();
```
Comment (jplafonta) [inline CppSource/Source/Services/Achievements/achievements_helper.cpp]: How does this get cleaned up?

File: CppSource/Source/Services/Achievements/achievements_helper.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+achievements_result XSAPI_ACHIEVEMENTS_RESULT_IMPL::cppAchievementsResult() const
+{
+    return m_cppResult;
+}
+
+XSAPI_ACHIEVEMENT* CreateAchievementFromCpp(
+    _In_ achievement cppAchievement
+)
+{
+    auto achievement = new XSAPI_ACHIEVEMENT();
```
Comment (jplafonta) [inline CppSource/Source/Services/Achievements/achievements_helper.cpp]: how does this get cleaned up?

File: CppSource/Include/xsapi/achievements_c.h (hunk 5 above)
Comment (rakhad) [reply CppSource/Include/xsapi/achievements_c.h]: AchievementService is the name of the c++ and winrt class. Those methods are the methods of that class.

File: CppSource/Source/Services/Achievements/achievements.cpp [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+using namespace xbox::services::achievements;
+
+xsapi_singleton* get_singleton_for_achievements()
+{
+    auto singleton = get_xsapi_singleton();
+
+    if (singleton->m_achievementsState == nullptr)
+    {
+        singleton->m_achievementsState = std::make_unique<XSAPI_ACHIEVEMENTS_STATE>();
+    }
```
Comment (jplafonta) [inline CppSource/Source/Services/Achievements/achievements.cpp]: you could just do this once in the xsapi_singleton ctor and then get rid of this function and use the existing get_xsapi_singleton

File: CppSource/Source/Services/Achievements/achievements.cpp [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    xbox_live_result<achievements_result> result = args->achievementsResult->pImpl->cppAchievementsResult().get_next(args->maxItems).get();
+    args->copy_xbox_live_result(result);
+
+    if (!result.err())
+    {
+        auto singleton = get_singleton_for_achievements();
+        std::lock_guard<std::recursive_mutex> lock(singleton->m_achievementsState->m_lock);
+
+        auto achievementsResult = CreateAchievementsResultFromCpp(result.payload());
+        singleton->m_achievementsState->m_achievementResults.insert(achievementsResult);
```
Comment (jplafonta) [inline CppSource/Source/Services/Achievements/achievements.cpp]: nit: instead of repeating this line add it to CreateAchievementsResultFromCpp

**Pull request merged**